File: ruby/tests/json_serialization_test.rb
require "json"
require "time"
require "securerandom"

class JSONSerializationTest
  attr_reader :name

  def initialize(obj_size = 5000, depth = 3, runs = 5)
    @name = "JSON Serialization"
    @obj_size = obj_size
    @depth = depth
    @runs = runs
  end

  def generate_nested(level)
    if level == 0
      {
        "id" => rand(1..1_000_000),
        "value" => rand,
        "active" => [true, false].sample,
        "text" => "x" * 50
      }
    else
      Hash[(0...@obj_size / 10).map { |i| ["child_#{i}", generate_nested(level - 1)] }]
    end
  end

  def run_once
    data = Array.new(20) { generate_nested(@depth) }

    t0 = Process.clock_gettime(Process::CLOCK_MONOTONIC)
    json_str = JSON.dump(data)
    t_mid = Process.clock_gettime(Process::CLOCK_MONOTONIC)
    JSON.parse(json_str)
    t1 = Process.clock_gettime(Process::CLOCK_MONOTONIC)

    encode_t = t_mid - t0
    decode_t = t1 - t_mid
    total = t1 - t0
    ops = json_str.size / (total > 0 ? total : 1e-9)

    {
      "encode_s" => encode_t,
      "decode_s" => decode_t,
      "total_s" => total,
      "ops_per_s" => ops
    }
  end

  def run
    results = Array.new(@runs) { run_once }
    totals = results.map { |r| r["total_s"] }.sort
    ops = results.map { |r| r["ops_per_s"] }.sort
    median_t = totals[totals.size / 2]
    median_ops = ops[ops.size / 2]

    {
      "name" => @name,
      "runs" => @runs,
      "median_total_s" => median_t,
      "median_ops_per_s" => median_ops,
      "raw" => results
    }
  end
end

if __FILE__ == $0
  result = JSONSerializationTest.new.run
  File.write("results_json_serialization_ruby.json", JSON.pretty_generate(result))
  puts JSON.pretty_generate(result)
end

File: python/tests/json_serialization_test.py
import json
import time
import random
import gc
import os

class JSONSerializationTest:
    def __init__(self, obj_size=5000, depth=3, runs=5, name="JSON Serialization"):
        self.name = name
        self.obj_size = obj_size
        self.depth = depth
        self.runs = runs

    def _generate_nested(self, level):
        if level == 0:
            return {
                "id": random.randint(1, 1_000_000),
                "value": random.random(),
                "active": random.choice([True, False]),
                "text": "x" * 50
            }
        else:
            return {
                f"child_{i}": self._generate_nested(level - 1)
                for i in range(self.obj_size // 10)
            }

    def run_once(self):
        # Generate nested JSON data
        data = [self._generate_nested(self.depth) for _ in range(20)]

        # Measure encode/decode performance
        t0 = time.perf_counter()
        s = json.dumps(data)
        mid = time.perf_counter()
        _ = json.loads(s)
        t1 = time.perf_counter()

        encode_t = mid - t0
        decode_t = t1 - mid
        total = t1 - t0
        ops = len(s) / total if total > 0 else 0

        return {
            "encode_s": encode_t,
            "decode_s": decode_t,
            "total_s": total,
            "ops_per_s": ops
        }

    def run(self, runs=None, iterations=None):
        # `iterations` unused but included for benchmark uniformity
        if runs is None:
            runs = self.runs

        results = []
        for _ in range(runs):
            results.append(self.run_once())
            gc.collect()

        totals = sorted(r["total_s"] for r in results)
        ops = sorted(r["ops_per_s"] for r in results)
        median_t = totals[len(totals)//2]
        median_ops = ops[len(ops)//2]

        return {
            "name": self.name,
            "runs": runs,
            "median_total_s": median_t,
            "median_ops_per_s": median_ops,
            "raw": results
        }

if __name__ == "__main__":
    os.makedirs("results", exist_ok=True)
    result = JSONSerializationTestTest().run()
    with open("results_json_serialization_python.json", "w") as f:
        json.dump(result, f, indent=2)
    print(f"{result['name']}: Median Time={result['median_total_s']:.4f}s | "
          f"Ops/sec={result['median_ops_per_s']:.2f}")
